File: FX/jahudka_consolefx_create_busses.lua
-- ReaScript Name: Create Airwindows Console post-fader busses for selected tracks
-- Version: 2.0
-- Author: jahudka
-- Links:
--   Documentation https://github.com/jahudka/reaper-consolefx
-- Changelog:
--   v2.0 (2020-12-21)
--    - extracted utility methods to *_utils.lua
--    - added support for Console 7 Cascade
--   v1.5 (2020-12-09)
--    - fixed Linked Gain Utility local path
--   v1.4 (2020-12-09)
--    - added support for Console 7
--    - add and configure Linked Gain Utility automatically
--   v1.3 (2020-02-16)
--    - don't create busses for tracks that don't have master / parent send
--   v1.2 (2020-01-05)
--    - set the pan law for created busses explicitly to +0.0 dB
--   v1.1 (2020-01-05)
--    - use VST on Windows automatically
--    - added all the Console types
--   v1.0 (2020-01-05)
--    - initial release
-- About:
--   This script can be used to automatically generate the appropriate routing
--   for the Airwindows Console effects, which need to be applied post-fader.
--   See https://github.com/jahudka/reaper-consolefx#creating-console-busses
--   for an usage overview.

-- HELPERS

package.path = (function ()
    local info = debug.getinfo(1, 'S')
    local s = package.config:sub(1, 1)
    return info.source:sub(2):gsub(s .. '[^' .. s .. ']+$', '') .. s .. '?.lua;' .. package.path
end)()

require('jahudka_consolefx_utils')

if not log then
    reaper.MB("This script depends on 'jahudka_consolefx_utils.lua' but it appears it isn't installed! Please install it first and then run this script again.", "Whoops!", 0)
end

local prefs = read_preferences()

local function create_bus(idx, name, mcp)
    reaper.InsertTrackAtIndex(idx, false)
    local track = reaper.GetTrack(0, idx)
    reaper.GetSetMediaTrackInfo_String(track, 'P_NAME', name, true)
    reaper.SetMediaTrackInfo_Value(track, 'B_SHOWINTCP', 0)
    reaper.SetMediaTrackInfo_Value(track, 'D_PANLAW', 1.0)

    if mcp ~= true then
        reaper.SetMediaTrackInfo_Value(track, 'B_SHOWINMIXER', 0)
    end

    return track
end

local function reroute(from_track, to_track)
    reaper.SetMediaTrackInfo_Value(from_track, 'B_MAINSEND', 0)
    local send = reaper.CreateTrackSend(from_track, to_track)
    reaper.SetTrackSendInfo_Value(from_track, 0, send, 'I_SENDMODE', 0)
end

local function get_free_bus_id()
    local id = 200

    for i = 1, reaper.CountTracks(0) do
        local track = reaper.GetTrack(0, i - 1)
        local parent = reaper.GetParentTrack(track)
        local parent_send = reaper.GetMediaTrackInfo_Value(track, 'B_MAINSEND')

        if is_console_track(track) and parent == nil and parent_send == 0 then
            id = id + 1
        end
    end

    return id
end

local function add_linked_gain_utility(track, link_id, invert)
    local id = reaper.TrackFX_AddByName(track, prefs['linked_gain_utility_path'], false, -1)

    if id > -1 then
        reaper.TrackFX_SetParam(track, id, 0, link_id)
        reaper.TrackFX_SetParam(track, id, 2, invert)
    end
end

local function add_console_fx(track, type)
    for _, t in ipairs(prefs['console_types']) do
        local id = reaper.TrackFX_AddByName(track, format_plugin_name(t, type), false, -1)

        if id > -1 and t ~= prefs['default_type'] then
            reaper.TrackFX_SetOffline(track, id, true)
        end
    end
end

-- END HELPERS

-- MAIN

local function create_console_busses()
    local n_selected = reaper.CountSelectedTracks(0)
    local parent_track
    local target_tracks

    if n_selected == 1 then
        parent_track = reaper.GetSelectedTrack(0, 0)
        target_tracks = get_child_tracks(parent_track)
    else
        target_tracks = get_selected_tracks()
    end

    if target_tracks == nil or #target_tracks < 1 then
        return
    end

    local idx = reaper.CountTracks(0)

    reaper.Undo_BeginBlock()

    local parent_name = parent_track and string.upper(get_track_name(parent_track)) or 'MASTER'
    local link_id = parent_track and get_free_bus_id() or 255
    local summing_bus = create_bus(idx, prefix_track_name(parent_name), true)

    reaper.SetMediaTrackInfo_Value(summing_bus, 'I_FOLDERDEPTH', 1)
    add_console_fx(summing_bus, 'Buss')
    add_linked_gain_utility(summing_bus, link_id, 1)

    if parent_track ~= nil then
        reroute(summing_bus, parent_track)
    end

    for _, track in ipairs(target_tracks) do
        if has_parent_send(track) then
            idx = idx + 1
            local bus = create_bus(idx, prefix_track_name(get_track_name(track)), false)
            add_linked_gain_utility(bus, link_id, 0)
            add_console_fx(bus, 'Channel')
            reroute(track, bus)
        end
    end

    local last_bus = reaper.GetTrack(0, idx)
    reaper.SetMediaTrackInfo_Value(last_bus, 'I_FOLDERDEPTH', -1)

    reaper.Undo_EndBlock('Create Airwindows Console post-fader busses for selected tracks', -1)
end

create_console_busses()

File: FX/jahudka_consolefx_create_busses.py
"""
ReaScript Name: Create Airwindows Console post-fader busses for selected tracks
About:
  Creates busses with Airwindows Console effects for the selected track(s).
  There are two modes of operation available: folder injection and global injection.

  ## Folder injection

  This mode inserts the Console plugins in between a folder track and its children.
  To use this mode, select a single folder track and run this script. If we call
  the selected folder track `F` and its children `T1` .. `Tn`, the following will happen:
   - A new folder track is created at the end of the project; it will be named `[C] F`.
   - For each of the children `Tn` of `F`, a new track called `[C] Tn` will be added to `[C] F`
   - All of the child tracks `Tn` of `F` will have their Master / parent send turned off,
     and instead each will have a post-fader send to its counterpart `[C] Tn` in `[C] F`
   - `[C] F` will also have its Master / parent send turned off and instead it will
     have a post-fader send to `F`
   - All configured Console Channel plugins will be inserted on all the `[C] Tn` tracks;
     all but the first plugin will be set offline
   - All configured Console Buss plugins will be inserted on `[C] F`; all but the first
     plugin will be set offline
   - All of the `[C] Tn` tracks will be hidden from both MCP and TCP
   - `[C] F` will be hidden from TCP

  ## Global injection

  This mode inserts the Console plugins in between all selected tracks and the master bus.
  To use this mode, select 2 or more tracks and run this script. In this mode the routing
  that is generated is similar to Folder injection; the only difference is that the summing
  bus (`[C] F` in the Folder injection example) would keep its Master / parent send on
  and wouldn't get a post-fader send.
Author: jahudka
Link: https://github.com/jahudka/reaper-consolefx
Version: 1.0
"""
from reaper_python import *
from jahudka_consolefx_common import *


def create_console_busses():
    if num_selected_tracks() < 1:
        return

    if num_selected_tracks() == 1:
        parent_track = RPR_GetSelectedTrack(0, 0)
        target_tracks = get_child_tracks(parent_track)

        if len(target_tracks) < 1:
            return
    else:
        parent_track = None
        target_tracks = get_selected_tracks()

    idx = num_all_tracks()

    RPR_Undo_BeginBlock()

    parent_name = get_track_name(parent_track).upper() if parent_track else 'MASTER'
    summing_bus = _create_bus(idx, '[C] ' + parent_name, True)
    RPR_SetMediaTrackInfo_Value(summing_bus, 'I_FOLDERDEPTH', 1)
    _add_console_fx(summing_bus, 'Buss')

    if parent_track:
        _reroute(summing_bus, parent_track)

    for track in target_tracks:
        idx += 1
        bus = _create_bus(idx, '[C] ' + get_track_name(track))
        _add_console_fx(bus, 'Channel')
        _reroute(track, bus)

    last_bus = RPR_GetTrack(0, idx)
    RPR_SetMediaTrackInfo_Value(last_bus, 'I_FOLDERDEPTH', -1)

    RPR_Undo_EndBlock("Create Airwindows Console post-fader busses for selected tracks", -1)


def _create_bus(idx, name, mcp=False):
    RPR_InsertTrackAtIndex(idx, False)
    track = RPR_GetTrack(0, idx)
    set_track_name(track, name)
    RPR_SetMediaTrackInfo_Value(track, 'B_SHOWINTCP', 0)

    if not mcp:
        RPR_SetMediaTrackInfo_Value(track, 'B_SHOWINMIXER', 0)

    return track


def _reroute(from_track, to_track):
    RPR_SetMediaTrackInfo_Value(from_track, 'B_MAINSEND', 0)
    send = RPR_CreateTrackSend(from_track, to_track)
    RPR_SetTrackSendInfo_Value(from_track, 0, send, 'I_SENDMODE', 0)


def _add_console_fx(track, plugin_type):
    idx = 0

    for console_type in CONSOLE_TYPES:
        RPR_TrackFX_AddByName(track, PLUGIN_NAME_FORMAT % (console_type, plugin_type), False, -1)

        if idx > 0:
            RPR_TrackFX_SetOffline(track, idx, True)

        idx += 1


create_console_busses()
